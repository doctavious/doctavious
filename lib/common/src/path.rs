use std::ffi::{OsStr, OsString};
use std::path::PathBuf;

pub fn append_to_path(p: impl Into<OsString>, s: impl AsRef<OsStr>) -> PathBuf {
    let mut p = p.into();
    p.push(s);
    p.into()
}
