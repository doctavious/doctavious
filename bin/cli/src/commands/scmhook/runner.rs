use std::path::PathBuf;
use std::process::Command;

use regex::{Error, Regex, RegexBuilder};
use scm::drivers::Scm;
use scm::hooks::{ScmHook, ScmHookConditionalExecution};
use tracing::warn;

// TODO: this should move to CLI

struct ScmRunnerOptions<'a> {
    hook: ScmHook,
    hook_name: String,
    force: bool,
    files: Vec<PathBuf>,
    run_only_commands: Vec<String>,
    scm: &'a Scm,
}

struct ScmRunner<'a> {
    pub options: ScmRunnerOptions<'a>,
}

impl<'a> ScmRunner<'a> {
    pub fn new(options: ScmRunnerOptions<'a>) -> Self {
        Self { options }
    }

    pub fn run_all(&self) {
        // this probably should look at run_only_commands as well?
        if self.should_skip(&self.options.hook) {
            return;
        }

        // run scripts
        // run commands
    }

    fn should_skip(&self, hook: &ScmHook) -> bool {
        ExecutionChecker::check(&hook.skip, &hook.only)
    }
}

struct ExecutionChecker;

impl ExecutionChecker {
    pub fn check(
        // &self,
        skip: &Option<ScmHookConditionalExecution>,
        only: &Option<ScmHookConditionalExecution>,
    ) -> bool {
        if let Some(skip) = skip {
            if Self::matches(skip) {
                return true;
            }
        }

        if let Some(only) = only {
            return !Self::matches(only);
        }

        false
    }

    fn matches(
        // &self,
        condition: &ScmHookConditionalExecution,
    ) -> bool {
        match condition {
            ScmHookConditionalExecution::Bool(b) => *b,
            ScmHookConditionalExecution::Ref(refs) => {
                for r in refs {
                    match RegexBuilder::new(&r).build() {
                        Ok(reg) => {
                            // if reg.is_match()
                        }
                        Err(e) => {
                            warn!("Invalid hook condition {r}: {e}")
                        }
                    }
                }
                todo!()
            }
            ScmHookConditionalExecution::Run(r) => {
                for s in r {
                    match s.split_once(" ") {
                        None => warn!("Invalid hook condition run command {s}"),
                        Some((cmd, args)) => {
                            match Command::new(cmd).args(args.split(" ")).output() {
                                Ok(output) => {
                                    // TODO: how to know if this resulted in true
                                    return true;
                                }
                                Err(e) => {
                                    warn!("Hook condition run command {s} failed: {e}")
                                }
                            }
                        }
                    }
                }
                todo!()
            }
            _ => {
                todo!()
            }
        }
    }
}
