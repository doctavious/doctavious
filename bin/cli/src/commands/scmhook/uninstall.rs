use std::path::PathBuf;

use clap::Parser;
use doctavious_cli::CliResult;

/// Clear SCM hooks related to doctavious configuration
#[derive(Parser, Debug)]
#[command()]
pub(crate) struct Uninstall {
    #[arg(long, short)]
    pub cwd: Option<PathBuf>,

    /// Flag to remove all SCM hooks even those not related to doctavious
    #[arg(long, short, action)]
    pub force: bool,

    /// Flag to remove SCM hook configuration from doctavious configuration
    #[arg(long, short, action)]
    pub remove_config: bool,
}

pub(crate) fn execute(command: Uninstall) -> CliResult<Option<String>> {
    Ok(None)
}
