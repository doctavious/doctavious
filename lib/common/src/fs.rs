use std::path::Path;
use std::{fs, io};
use std::io::{Error, ErrorKind};

// TODO: I feel like I could do this better
pub fn copy_dir(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> io::Result<()> {
    if !src.as_ref().is_dir() {
        // TODO (sean): change to ErrorKind::NotADirectory when io_error_more stabilizes
        return Err(Error::new(ErrorKind::Other, "Not a directory"));
    }

    let dest = if src.as_ref().to_string_lossy().ends_with("/") {
        dst.as_ref().to_path_buf()
    } else {
        dst.as_ref().join(src.as_ref().file_name().unwrap())
    };

    recursive_dir_copy(src, dst)?;

    Ok(())
}

fn recursive_dir_copy(src: impl AsRef<Path>, dst: impl AsRef<Path>) -> io::Result<()> {
    fs::create_dir_all(&dst)?;
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        if entry.file_type()?.is_dir() {
            recursive_dir_copy(entry.path(), dst.as_ref().join(entry.file_name()))?;
        } else {
            fs::copy(entry.path(), dst.as_ref().join(entry.file_name()))?;
        }
    }
    Ok(())
}
