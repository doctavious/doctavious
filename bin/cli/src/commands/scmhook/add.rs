use std::path::PathBuf;

use clap::Parser;
use doctavious_cli::cmd::scm_hooks::add::add;
use doctavious_cli::CliResult;

/// Create a SCM Hook.
///
/// Similar to what `scmhook install` command does but doesn't create a configuration first.
#[derive(Parser, Debug)]
#[command()]
pub(crate) struct AddScmHook {
    #[arg(long, short)]
    pub cwd: Option<PathBuf>,

    /// SCM Hook name
    #[arg(index = 1)]
    pub name: String,

    /// Whether to create a directory for scripts
    #[arg(long, short, action)]
    pub dir: bool,

    /// Overwrite .old hooks
    #[arg(long, short, action)]
    pub force: bool,
}

pub(crate) fn execute(command: AddScmHook) -> CliResult<Option<String>> {
    let path = command.cwd.unwrap_or(std::env::current_dir()?);

    add(&path, command.name, command.dir, command.force)?;

    Ok(None)
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::PathBuf;

    use common::fs::copy_dir;
    use scm::drivers::git::GitScmRepository;
    use scm::hooks::OLD_HOOK_POSTFIX;
    use scm::{ScmRepository, HOOK_TEMPLATE};
    use tempfile::TempDir;
    use testing::cleanup::CleanUp;

    use crate::commands::scmhook::add::{execute, AddScmHook};

    #[test]
    fn should_fail_if_scm_not_initialized() {
        let temp_dir = TempDir::new().unwrap();
        let temp_path = temp_dir.into_path();
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: false,
            force: false,
        });

        assert!(result.is_err());
        assert_eq!(
            "SCM error: Could not find supported SCM",
            result.unwrap_err().to_string()
        );
    }

    #[test]
    fn should_fail_with_invalid_hook() {
        let (temp_path, _) = setup(None);
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "invalid-hook".to_string(),
            dir: false,
            force: false,
        });

        assert!(result.is_err());
        assert_eq!(
            "SCM error: Hook invalid-hook is not supported",
            result.unwrap_err().to_string()
        );
    }

    #[test]
    fn should_add_without_doctavious_configuration() {
        let (temp_path, scm) = setup(None);
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: false,
            force: false,
        });

        assert!(result.is_ok());
        assert!(scm.hooks_path().unwrap().join("pre-commit").exists());
        assert!(!temp_path.join(".doctavious/scmhooks/pre-commit").exists());
    }

    #[test]
    fn should_add_with_doctavious_configured() {
        let config = r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit.executions.format-backend]
type = "command"
run = "cargo fmt"
root = "backend"
"###;

        let (temp_path, scm) = setup(Some(config));
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: false,
            force: false,
        });

        assert!(result.is_ok());
        let pre_commit_path = scm.hooks_path().unwrap().join("pre-commit");
        assert!(pre_commit_path.exists());
        assert!(fs::read_to_string(pre_commit_path)
            .unwrap()
            .contains("doctavious"));
    }

    #[test]
    fn should_create_hooks_script_configuration_directory() {
        let (temp_path, scm) = setup(None);
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: true,
            force: false,
        });

        assert!(result.is_ok());
        assert!(scm.hooks_path().unwrap().join("pre-commit").exists());
        assert!(temp_path.join(".doctavious/scmhooks/pre-commit").is_dir());
    }

    #[test]
    fn should_replace_existing_hook() {
        let (temp_path, scm) = setup(None);
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        fs::write(&pre_commit_path, "some hook content").unwrap();

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: false,
            force: false,
        });

        assert!(result.is_ok());
        assert!(pre_commit_path.exists());
        assert!(fs::read_to_string(pre_commit_path)
            .unwrap()
            .contains("doctavious"));
        assert!(scm_hooks_path.join("pre-commit.old").exists());
    }

    #[test]
    fn should_replace_existing_doctavious_hook() {
        let (temp_path, scm) = setup(None);
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        fs::write(&pre_commit_path, HOOK_TEMPLATE).unwrap();

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: false,
            force: false,
        });

        assert!(result.is_ok());
        assert!(pre_commit_path.exists());
        assert!(fs::read_to_string(pre_commit_path)
            .unwrap()
            .contains("doctavious"));
        assert!(!scm_hooks_path.join("pre-commit.old").exists());
    }

    #[test]
    fn should_fail_with_existing_old_hook() {
        let (temp_path, scm) = setup(None);
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        let old_pre_commit_path = common::path::append_to_path(&pre_commit_path, OLD_HOOK_POSTFIX);
        fs::write(&pre_commit_path, "some hook content").unwrap();
        fs::write(&old_pre_commit_path, "some old hook content").unwrap();

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: false,
            force: false,
        });

        assert!(result.is_err());
        assert_eq!(
            "SCM error: Can't rename pre-commit to pre-commit.old as file already exists. If you wish to overwrite use 'force' option",
            result.unwrap_err().to_string()
        );

        let hooks_path = scm.hooks_path().unwrap();
        assert!(hooks_path.join("pre-commit").exists());
        assert!(hooks_path.join("pre-commit.old").exists());
    }

    #[test]
    fn should_replace_hook_and_overwrite_old_with_force() {
        let (temp_path, scm) = setup(None);
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        let old_pre_commit_path = common::path::append_to_path(&pre_commit_path, OLD_HOOK_POSTFIX);
        fs::write(&pre_commit_path, "some hook content").unwrap();
        fs::write(&old_pre_commit_path, "some old hook content").unwrap();

        let result = execute(AddScmHook {
            cwd: Some(temp_path.clone()),
            name: "pre-commit".to_string(),
            dir: false,
            force: true,
        });

        assert!(result.is_ok());
        assert!(pre_commit_path.exists());
        assert!(fs::read_to_string(pre_commit_path)
            .unwrap()
            .contains("doctavious"));
        assert_eq!(
            "some hook content",
            fs::read_to_string(&old_pre_commit_path).unwrap()
        );
    }

    fn setup(doctavous_config: Option<&str>) -> (PathBuf, GitScmRepository) {
        let temp_dir = TempDir::new().unwrap();
        let temp_path = temp_dir.into_path();

        copy_dir("./tests/fixtures/scmhook/", &temp_path).expect("copy test fixtures");
        if let Some(config) = doctavous_config {
            fs::write(temp_path.join("doctavious.toml"), config).expect("write doctavious.toml");
        }

        let scm = GitScmRepository::init(&temp_path).expect("init git");
        scm.add_all();

        (temp_path, scm)
    }
}
