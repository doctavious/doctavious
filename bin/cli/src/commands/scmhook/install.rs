use std::path::PathBuf;

use clap::Parser;
use doctavious_cli::cmd::scm_hooks::install::install;

/// Synchronize SCM hooks with your configuration.
#[derive(Parser, Debug)]
#[command()]
pub struct InstallScmHook {
    #[arg(long, short)]
    pub cwd: Option<PathBuf>,

    /// Overwrite .old hooks
    #[arg(long, short, action)]
    pub force: bool,
}

#[async_trait::async_trait]
impl crate::commands::Command for InstallScmHook {
    async fn execute(&self) -> anyhow::Result<Option<String>> {
        let cwd = self.resolve_cwd(self.cwd.as_ref())?;
        install(&cwd, self.force)?;
        Ok(None)
    }
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::PathBuf;

    use doctavious_cli::cmd::scm_hooks::{HOOK_TEMPLATE, HOOK_TEMPLATE_CHECKSUM};
    use doctavious_cli::settings::Config;
    use doctavious_std::fs::copy_dir;
    use scm::drivers::ScmRepository;
    use scm::drivers::git::GitScmRepository;
    use scm::hooks::OLD_HOOK_POSTFIX;
    use tempfile::TempDir;
    use testing::cleanup::CleanUp;

    use crate::commands::Command;
    use crate::commands::scmhook::install::InstallScmHook;

    // TODO: should_install_without_doctavious_config

    #[tokio::test]
    async fn should_install_with_existing_doctavious_config() {
        let config = r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit.executions.format-backend]
type = "command"
run = "cargo fmt"
root = "backend"
[scmhook_settings.hooks.post-commit.executions.echo]
type = "command"
run = "echo 'Done!'"
"###;

        let (temp_path, scm) = setup(Some(config));
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let cmd = InstallScmHook {
            cwd: Some(temp_path.clone()),
            force: false,
        };

        let result = cmd.execute().await;

        let hooks_path = scm.hooks_path().unwrap();
        assert!(result.is_ok());
        assert!(hooks_path.join("pre-commit").exists());
        assert!(hooks_path.join("post-commit").exists());

        let synchronization_file = scm.info_path().unwrap().join("doctavious.synchronization");
        assert!(synchronization_file.exists());
        assert_eq!(
            format!(
                "{}|ae6e1c36d1f298f4692eed15be33a2ae",
                HOOK_TEMPLATE_CHECKSUM.as_str()
            ),
            fs::read_to_string(synchronization_file).unwrap()
        );
    }

    #[tokio::test]
    async fn should_install_with_existing_hooks() {
        let config = r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit.executions.format-backend]
type = "command"
run = "cargo fmt"
root = "backend"
[scmhook_settings.hooks.post-commit.executions.echo]
type = "command"
run = "echo 'Done!'"
"###;

        let (temp_path, scm) = setup(Some(config));
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        fs::write(&pre_commit_path, "some hook content").unwrap();

        let cmd = InstallScmHook {
            cwd: Some(temp_path.clone()),
            force: false,
        };

        let result = cmd.execute().await;

        let hooks_path = scm.hooks_path().unwrap();
        assert!(result.is_ok());
        assert!(hooks_path.join("pre-commit").exists());
        assert!(
            fs::read_to_string(&pre_commit_path)
                .unwrap()
                .contains("doctavious")
        );

        let old_hook_path = hooks_path.join("pre-commit.old");
        assert!(old_hook_path.exists());
        assert_eq!(
            "some hook content",
            fs::read_to_string(&old_hook_path).unwrap()
        );
        assert!(hooks_path.join("post-commit").exists());
        assert!(
            scm.info_path()
                .unwrap()
                .join("doctavious.synchronization")
                .exists()
        );
    }

    #[tokio::test]
    async fn should_install_with_existing_doctavious_hooks() {
        let config = r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit.executions.format-backend]
type = "command"
run = "cargo fmt"
root = "backend"
[scmhook_settings.hooks.post-commit.executions.echo]
type = "command"
run = "echo 'Done!'"
"###;

        let (temp_path, scm) = setup(Some(config));
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        fs::write(&pre_commit_path, HOOK_TEMPLATE).unwrap();

        let cmd = InstallScmHook {
            cwd: Some(temp_path.clone()),
            force: false,
        };

        let result = cmd.execute().await;

        let hooks_path = scm.hooks_path().unwrap();
        assert!(result.is_ok());
        assert!(hooks_path.join("pre-commit").exists());
        assert!(
            fs::read_to_string(&pre_commit_path)
                .unwrap()
                .contains("doctavious")
        );

        assert!(!hooks_path.join("pre-commit.old").exists());
    }

    #[tokio::test]
    async fn should_update_stale_synchronization_file() {
        for (hook_template_checksum, settings_checksum) in vec![
            ("123abc", "ae6e1c36d1f298f4692eed15be33a2ae"), // stale hook template checksum
            (HOOK_TEMPLATE_CHECKSUM.as_str(), "123abc"),    // stale settings checksum
        ] {
            let config = r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit.executions.format-backend]
type = "command"
run = "cargo fmt"
root = "backend"
[scmhook_settings.hooks.post-commit.executions.echo]
type = "command"
run = "echo 'Done!'"
"###;

            let (temp_path, scm) = setup(Some(config));
            let _c = CleanUp::new(Box::new(|| {
                let _ = fs::remove_dir_all(&temp_path);
            }));

            let scm_hooks_path = scm.hooks_path().unwrap();
            let pre_commit_path = scm_hooks_path.join("pre-commit");
            fs::write(&pre_commit_path, HOOK_TEMPLATE).unwrap();

            let synchronization_file = scm.info_path().unwrap().join("doctavious.synchronization");
            fs::write(&synchronization_file, "123|").unwrap();

            let cmd = InstallScmHook {
                cwd: Some(temp_path.clone()),
                force: false,
            };

            let result = cmd.execute().await;

            let hooks_path = scm.hooks_path().unwrap();
            assert!(result.is_ok());
            assert!(hooks_path.join("pre-commit").exists());
            assert!(
                fs::read_to_string(&pre_commit_path)
                    .unwrap()
                    .contains("doctavious")
            );

            assert!(!hooks_path.join("pre-commit.old").exists());
            assert!(hooks_path.join("post-commit").exists());

            assert_eq!(
                format!(
                    "{}|ae6e1c36d1f298f4692eed15be33a2ae",
                    HOOK_TEMPLATE_CHECKSUM.as_str()
                ),
                fs::read_to_string(synchronization_file).unwrap()
            );
        }
    }

    #[tokio::test]
    async fn should_fail_with_existing_hook_and_old() {
        let config = r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit.executions.format-backend]
type = "command"
run = "cargo fmt"
root = "backend"
[scmhook_settings.hooks.post-commit.executions.echo]
type = "command"
run = "echo 'Done!'"
"###;

        let (temp_path, scm) = setup(Some(config));
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        let old_pre_commit_path =
            doctavious_std::path::append_to_path(&pre_commit_path, OLD_HOOK_POSTFIX);
        fs::write(&pre_commit_path, "some hook content").unwrap();
        fs::write(&old_pre_commit_path, "some old hook content").unwrap();

        let cmd = InstallScmHook {
            cwd: Some(temp_path.clone()),
            force: false,
        };

        let result = cmd.execute().await;

        assert!(result.is_err());
        assert_eq!(
            "SCM error: Can't rename pre-commit to pre-commit.old as file already exists. If you wish to overwrite use 'force' option",
            result.unwrap_err().to_string()
        );

        let hooks_path = scm.hooks_path().unwrap();
        assert!(hooks_path.join("pre-commit").exists());
        assert!(hooks_path.join("pre-commit.old").exists());
        assert!(!hooks_path.join("post-commit").exists());
        assert!(
            !scm.info_path()
                .unwrap()
                .join("doctavious.synchronization")
                .exists()
        );
    }

    #[tokio::test]
    async fn should_install_with_existing_hook_and_old_with_force() {
        let config = r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit.executions.format-backend]
type = "command"
run = "cargo fmt"
root = "backend"
[scmhook_settings.hooks.post-commit.executions.echo]
type = "command"
run = "echo 'Done!'"
"###;

        let (temp_path, scm) = setup(Some(config));
        let _c = CleanUp::new(Box::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        }));

        let scm_hooks_path = scm.hooks_path().unwrap();
        let pre_commit_path = scm_hooks_path.join("pre-commit");
        let old_pre_commit_path =
            doctavious_std::path::append_to_path(&pre_commit_path, OLD_HOOK_POSTFIX);
        fs::write(&pre_commit_path, "some hook content").unwrap();
        fs::write(&old_pre_commit_path, "some old hook content").unwrap();

        let cmd = InstallScmHook {
            cwd: Some(temp_path.clone()),
            force: true,
        };

        let result = cmd.execute().await;

        assert!(result.is_ok());

        let hooks_path = scm.hooks_path().unwrap();
        assert!(hooks_path.join("pre-commit").exists());
        assert!(
            fs::read_to_string(pre_commit_path)
                .unwrap()
                .contains("doctavious")
        );

        assert!(hooks_path.join("pre-commit.old").exists());
        assert_eq!(
            "some hook content",
            fs::read_to_string(old_pre_commit_path).unwrap()
        );

        let post_commit_path = hooks_path.join("post-commit");
        assert!(post_commit_path.exists());
        assert!(
            fs::read_to_string(post_commit_path)
                .unwrap()
                .contains("doctavious")
        );
        assert!(
            scm.info_path()
                .unwrap()
                .join("doctavious.synchronization")
                .exists()
        );
    }

    fn setup(doctavous_config: Option<&str>) -> (PathBuf, GitScmRepository) {
        let temp_dir = TempDir::new().unwrap();
        let temp_path = temp_dir.keep();

        copy_dir("./tests/fixtures/scmhook/", &temp_path).expect("copy test fixtures");
        if let Some(config) = doctavous_config {
            fs::write(temp_path.join(Config::config_file_path()), config)
                .expect("write doctavious.toml");
        }

        let scm = GitScmRepository::init(&temp_path).expect("init git");
        scm.add_all().expect("Should add all files to SCM");

        (temp_path, scm)
    }
}
