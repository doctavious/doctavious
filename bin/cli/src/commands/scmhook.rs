use clap::Parser;

use crate::commands::scmhook::add::AddScmHook;
use crate::commands::scmhook::install::InstallScmHook;
use crate::commands::scmhook::run::RunScmHookCommand;
use crate::commands::scmhook::uninstall::UninstallScmHook;

mod add;
mod install;
mod run;
mod uninstall;

/// Manage SCM Hooks
#[derive(Parser, Debug)]
#[command()]
pub struct ScmHookCommand {
    #[command(subcommand)]
    pub sub_command: ScmHookSubCommand,
}

#[remain::sorted]
#[derive(Parser, Debug)]
pub enum ScmHookSubCommand {
    Add(AddScmHook),
    Install(InstallScmHook),
    Run(RunScmHookCommand),
    Uninstall(UninstallScmHook),
}

#[async_trait::async_trait]
impl crate::commands::Command for ScmHookCommand {
    async fn execute(&self) -> anyhow::Result<Option<String>> {
        match &self.sub_command {
            ScmHookSubCommand::Add(cmd) => cmd.execute().await,
            ScmHookSubCommand::Install(cmd) => cmd.execute().await,
            ScmHookSubCommand::Run(cmd) => cmd.execute().await,
            ScmHookSubCommand::Uninstall(cmd) => cmd.execute().await,
        }
    }
}
