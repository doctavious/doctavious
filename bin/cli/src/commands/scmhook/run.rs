use std::path::PathBuf;

use clap::Parser;
use doctavious_cli::cmd::scm_hooks::run::{run, ScmHookRunFiles};
use doctavious_cli::CliResult;

/// Execute commands/scripts associated to the specified hook.
///
/// This is called for every hook managed by Doctavious.
/// You can also provide your own hooks that can only be called manually.
#[derive(Parser, Debug)]
#[command()]
pub(crate) struct RunScmHookCommand {
    /// Name of the hook to run
    #[arg(index = 1)]
    pub hook: String,

    /// Path to execute run
    #[arg(long, short)]
    pub cwd: Option<PathBuf>,

    // TODO: better explanation as to what files mean. Explain {files}
    // TODO: can use group = "files" to only allow one to be used?
    /// Run on specified file (repeat for multiple files). takes precedence over --all-files
    #[arg(long, short)]
    pub file: Option<Vec<PathBuf>>,

    // TODO: better explanation as to what files mean. Explain {files}
    /// Run hooks on all files
    #[arg(long, short, action)]
    pub all_files: Option<bool>,

    /// Run only specified executions (commands / scripts)
    #[arg(long = "executions", short = 'e')]
    pub run_only_executions: Option<Vec<String>>,

    /// Force execution of commands that can be skipped
    #[arg(long, short, action)]
    pub force: bool,
}

pub(crate) fn execute(command: RunScmHookCommand) -> CliResult<Option<String>> {
    let path = command.cwd.unwrap_or(std::env::current_dir()?);

    let cmd_files = command.file.unwrap_or_default();
    let files = if !cmd_files.is_empty() {
        Some(ScmHookRunFiles::Specific(cmd_files))
    } else if let Some(true) = command.all_files {
        Some(ScmHookRunFiles::All)
    } else {
        None
    };

    run(
        &path,
        &command.hook,
        files,
        command.run_only_executions.unwrap_or_default(),
        command.force,
    )?;

    Ok(None)
}

#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::PathBuf;

    use common::fs::copy_dir;
    use doctavious_cli::CliResult;
    use scm::drivers::git::GitScmRepository;
    use tempfile::TempDir;
    use testing::CleanUp;

    use crate::commands::scmhook::run::{execute, RunScmHookCommand};

    #[test]
    fn execute_hook() {
        let temp_dir = TempDir::new().unwrap();
        let temp_path = temp_dir.into_path();

        let c = CleanUp::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        });

        copy_dir("./tests/fixtures/scmhook/", &temp_path).expect("copy test fixtures");
        fs::write(
            temp_path.join("doctavious.toml"),
            r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit]
name = "pre-commit"
[scmhook_settings.hooks.pre-commit.executions.format-backend]
name = "format-backend"
type = "command"
run = "cargo fmt"
root = "backend"
tags = ["backed", "style"]
"###,
        )
            .expect("write doctavious.toml");

        let scm = GitScmRepository::init(&temp_path).expect("init git");
        scm.add_all();

        // std::process::Command::new("pnpm").arg("install").output().expect("pnpm install");

        let result = execute(RunScmHookCommand {
            hook: "pre-commit".to_string(),
            cwd: Some(temp_path.clone()),
            file: None,
            all_files: None,
            run_only_executions: None,
            force: false,
        });

        assert!(result.is_ok());
        insta::assert_snapshot!(fs::read_to_string(&temp_path.join("backend/src/lib.rs")).unwrap());
    }

    #[test]
    fn specified_files() {
        let temp_dir = TempDir::new().unwrap();
        let temp_path = temp_dir.into_path();

        let c = CleanUp::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        });

        copy_dir("./tests/fixtures/scmhook/", &temp_path).expect("copy test fixtures");
        fs::write(
            temp_path.join("doctavious.toml"),
            r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit]
name = "pre-commit"
[scmhook_settings.hooks.pre-commit.executions.specified-files]
name = "specified-files"
type = "command"
run = "echo '{files}' > test_specified_files.txt"
"###,
        )
        .expect("write doctavious.toml");

        let scm = GitScmRepository::init(&temp_path).expect("init git");
        scm.add_all();

        let result = execute(RunScmHookCommand {
            hook: "pre-commit".to_string(),
            cwd: Some(temp_path.clone()),
            file: Some(vec![PathBuf::from("/backend/src/lib.rs")]),
            all_files: None,
            run_only_executions: None,
            force: false,
        });

        assert!(result.is_ok());
        insta::assert_snapshot!(fs::read_to_string(&temp_path.join("test_specified_files.txt")).unwrap());
    }

    #[test]
    fn all_files() {
        let temp_dir = TempDir::new().unwrap();
        let temp_path = temp_dir.into_path();

        let c = CleanUp::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        });

        copy_dir("./tests/fixtures/scmhook/", &temp_path).expect("copy test fixtures");
        fs::write(
            temp_path.join("doctavious.toml"),
            r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit]
name = "pre-commit"
[scmhook_settings.hooks.pre-commit.executions.all-files]
name = "all-files"
type = "command"
run = "echo '{files}' > test_all_files.txt"
"###,
        )
            .expect("write doctavious.toml");

        let scm = GitScmRepository::init(&temp_path).expect("init git");
        scm.add_all();

        let result = execute(RunScmHookCommand {
            hook: "pre-commit".to_string(),
            cwd: Some(temp_path.clone()),
            file: None,
            all_files: Some(true),
            run_only_executions: None,
            force: false,
        });

        assert!(result.is_ok());
        insta::assert_snapshot!(fs::read_to_string(&temp_path.join("test_all_files.txt")).unwrap());
    }

    #[test]
    fn script() {
        let temp_dir = TempDir::new().unwrap();
        let temp_path = temp_dir.into_path();

        let c = CleanUp::new(|| {
            let _ = fs::remove_dir_all(&temp_path);
        });

        copy_dir("./tests/fixtures/scmhook/", &temp_path).expect("copy test fixtures");
        fs::write(
            temp_path.join("doctavious.toml"),
            r###"[scmhook_settings]
[scmhook_settings.hooks.pre-commit]
name = "pre-commit"
[scmhook_settings.hooks.pre-commit.executions.good-script]
file_name = "good-script.sh"
type = "script"
runner = "bash"
"###,
        )
            .expect("write doctavious.toml");

        let scm = GitScmRepository::init(&temp_path).expect("init git");
        scm.add_all();

        let result = execute(RunScmHookCommand {
            hook: "pre-commit".to_string(),
            cwd: Some(temp_path.clone()),
            file: None,
            all_files: None,
            run_only_executions: None,
            force: false,
        });

        assert!(result.is_ok());
        insta::assert_snapshot!(fs::read_to_string(&temp_path.join("script_output.txt")).unwrap());
    }

    // test run_only_executions

    // test tags

    // test force

    // test script
}
