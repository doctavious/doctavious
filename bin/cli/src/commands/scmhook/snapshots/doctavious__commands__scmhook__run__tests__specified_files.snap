---
source: bin/cli/src/commands/scmhook/run.rs
expression: "fs::read_to_string(&temp_path.join(\"test_specified_files.txt\")).unwrap()"
---
/backend/src/lib.rs

